### PR TITLE
Fix GroupBy nth Handling with Observed=False

### LINKED ISSUE #26385: GroupBy nth ignores observed keyword for Categorical
Was looking to consolidate nth, first and last in the GroupBy world but was hitting some issues with categorical tests and handling of missing data. Note that the following calls are not equivalent, though they should be:

```python
>>> import numpy as np
>>> import pandas as pd

>>> s1 = pd.Categorical([np.nan, 'a', np.nan, 'a'], categories=['a', 'b', 'c'])
>>> s2 = pd.Series([1, 2, 3, 4])
>>> df = pd.DataFrame({'s1': s1, 's2': s2})
>>> grp = df.groupby('s1', observed=False)
>>> grp.first()
     s2
s1
a   2.0
b   NaN
c   NaN

>>> grp.nth(0)
    s2
s1
a    2
```

### PR DESCRIPTION
**Pull request by WillAyd**
- [X] closes #26385
- [X] tests added / passed
- [] passes `git diff upstream/master -u -- "*.py" | flake8 --diff`
- [X] whatsnew entry

### CODE REVIEW

File: pandas/core/groupby/groupby.py
Comment (WillAyd) [inline pandas/core/groupby/groupby.py]: I was hoping to get this into `_wrap_aggregated_output` (which is what wraps most of the other methods) but that method doesn't appear to do any reindexing so would have to mess with that to make it work. Figured this as a one-off was easier though ideally would have everything consolidated...

File: pandas/tests/groupby/test_categorical.py
Comment (WillAyd) [inline pandas/tests/groupby/test_categorical.py]: Seemed strange that sum didn't return NA values where as the others did. Prod definitely seems off as well - I think these are existing bugs but can check in more detail in a few days

File: pandas/tests/groupby/conftest.py
Comment (WillAyd) [inline pandas/tests/groupby/conftest.py]: While not specific to this PR I figured this was a good time to introduce this and increase / standardize coverage

File: pandas/core/groupby/groupby.py
Comment (WillAyd) [inline pandas/core/groupby/groupby.py]: Also note that this reindexing only works when the user does not provided a `dropna` argument to `nth`, but that seemed to cause an orthogonal issue I can open later

File: pandas/core/groupby/groupby.py
Comment (jreback) [reply pandas/core/groupby/groupby.py]: you just need a small change I think in ``result_index`. This is very similar to what we do with ``all_grouper``.

**Review by jreback — CHANGES_REQUESTED** [review]

File: pandas/tests/groupby/test_categorical.py
Comment (jschendel) [reply pandas/tests/groupby/test_categorical.py]: This looks to be consistent with one of the [changes described in 0.22.0](http://pandas.pydata.org/pandas-docs/stable/whatsnew/v0.22.0.html#grouping-by-a-categorical).

File: pandas/core/groupby/groupby.py
Comment (WillAyd) [reply pandas/core/groupby/groupby.py]: IIUC I think `result_index` is fine and handles properly; the problem here is specific to `nth` where the mask on the preceding lines essentially ignores the work that `result_index` does to properly handle NA categorical data, hence the reindex

File: pandas/tests/groupby/test_categorical.py
Comment (WillAyd) [reply pandas/tests/groupby/test_categorical.py]: Great find. I guess this is intentional then just to support sum - cc @TomAugspurger if you have any insights

Comment (TomAugspurger) [reply pandas/tests/groupby/test_categorical.py]: I haven't looked at the context, but @jschendel is correct that we define `sum([])` to be 0 and `prod([])` to be 1.

Comment (WillAyd) [reply pandas/tests/groupby/test_categorical.py]: Good to know - I'll fix some of the comments up on next iteration

File: pandas/tests/groupby/conftest.py
Comment (jreback) [inline pandas/tests/groupby/conftest.py]: we have these defined in pandas/contest.py, all numeric reductions; add something similiar, but you can use groupby in the name, the reason is that this can then easily be shared with resample tests for example.

also need a description. and a more descriptive name.

File: pandas/tests/groupby/test_categorical.py
Comment (jreback) [inline pandas/tests/groupby/test_categorical.py]: issue number

File: pandas/tests/groupby/test_categorical.py
Comment (jreback) [inline pandas/tests/groupby/test_categorical.py]: needs to test using the observed fixture (meaning both True and False)

File: pandas/tests/groupby/test_categorical.py
Comment (jreback) [inline pandas/tests/groupby/test_categorical.py]: I would do nth separately, its not similar to the other reduction functions, and the test becomes much more straightforward.

File: pandas/core/groupby/groupby.py
Comment (jreback) [reply pandas/core/groupby/groupby.py]: I am not convinced here; this looks like just a bandaid to me and there is an underlying issue that needs fixing. also what about dropna=True?

**Review by jreback — CHANGES_REQUESTED** [review]

File: pandas/tests/groupby/conftest.py
Comment (WillAyd) [reply pandas/tests/groupby/conftest.py]: Make sense, though given feedback on doing `nth` separately below I might just defer this for another PR

File: pandas/core/groupby/groupby.py
Comment (WillAyd) [reply pandas/core/groupby/groupby.py]: `dropna='all'` seems to have a separate set of issues. Here's behavior on master:

```python
>>> grp.nth(0, dropna='all')
cat
a    1.0
b    2.0
c    3.0
Name: ser, dtype: float64
```
Which is not correct since `b` and `c` are never associated with those values. 

Also trying to mix that with observed doesn't work:

```python-traceback
>>> grp = df.groupby('cat', observed=True)['ser']
>>> grp.nth(0, dropna='all')
Traceback (most recent call last):
  File "<stdin>", line 1, in <module>
  File "/Users/williamayd/clones/pandas/pandas/core/groupby/groupby.py", line 1773, in nth
    result.index = self.grouper.result_index
  File "/Users/williamayd/clones/pandas/pandas/core/generic.py", line 5144, in __setattr__
    return object.__setattr__(self, name, value)
  File "pandas/_libs/properties.pyx", line 67, in pandas._libs.properties.AxisProperty.__set__
    obj._set_axis(self.axis, value)
  File "/Users/williamayd/clones/pandas/pandas/core/series.py", line 381, in _set_axis
    self._data.set_axis(axis, labels)
  File "/Users/williamayd/clones/pandas/pandas/core/internals/managers.py", line 155, in set_axis
    'values have {new} elements'.format(old=old_len, new=new_len))
ValueError: Length mismatch: Expected axis has 3 elements, new values have 1 elements
```

So both weird though I'm not sure how we want to handle the combinations of observed and dropna; will open a separate issue

Comment (WillAyd) [reply pandas/core/groupby/groupby.py]: See #26454

**Comment by WillAyd** [comment]
I think this makes sense to wait until #26463 gets merged and use `_reindex_like` from that once it gets moved into the base GroupBy class

**Comment by WillAyd** [comment]
So doesn't look like merging in those changes helped; I think we have a few random reindexing ops scattered throughout the module still investigating better ways to share that functionality

**Comment by jreback** [comment]
what's the status here?

**Comment by WillAyd** [comment]
Haven't been able to address this comment yet https://github.com/pandas-dev/pandas/pull/26419/files#r285344403 so depending on how much of a blocker you think that is may or may not be ready for merge.

I think as is gets us a lot closer to unifying nth / first / last so ultimately net benefit but open to debate

**Comment by TomAugspurger** [comment]
@WillAyd linting issue: https://dev.azure.com/pandas-dev/pandas/_build/results?buildId=14466&view=logs&jobId=435ca956-9126-505f-566f-fff31072e2ba&taskId=ffc59b44-7438-5626-aa7a-09fa4141874b&lineStart=43&lineEnd=44&colStart=1&colEnd=1

**Comment by jreback** [comment]
@WillAyd what was that comment above? this PR looks good.

**Comment by WillAyd** [comment]
https://github.com/pandas-dev/pandas/pull/26419/commits/ee549ed668d485d342c5ed9589eaea8327bb7401#r285344403

**Comment by WillAyd** [comment]
Hmm link to comments in old commits might not work but if you look at the entire commit you'll see it as the first one [there](https://github.com/pandas-dev/pandas/pull/26419/commits/ee549ed668d485d342c5ed9589eaea8327bb7401#r285344403)

File: doc/source/whatsnew/v0.25.0.rst
Comment (jreback) [inline doc/source/whatsnew/v0.25.0.rst]: move to 0.25.1

**Review by jreback — CHANGES_REQUESTED** [review]
lgtm. ping on green.

**Comment by TomAugspurger** [comment]
Fixed the merge conflict.

**Comment by TomAugspurger** [comment]
Thanks @WillAyd. I think this is an improvement on master.

I believe there's an outstanding request for additional test coverage with `observed=True`? Can you double check that when you have time, and perhaps open an issue for it?

**Comment by WillAyd** [comment]
Thanks @TomAugspurger for pushing this over the finish line

**Pull request merged**
